micropython/lora: Add stm32wl5 modem driver

projectgus opened the pull request:
Adds a LoRa driver for the STM32WL5 "SUBGHZ" LoRa modem, which is essentially a combined SX1261 & SX1262 attached to a custom interface inside the SoC.

This PR depends on Micropython built with https://github.com/micropython/micropython/pull/9130 for the hardware level support.

## TODOs

* [x] Update the package structure once the linked PR has one!
* ~~Fix HP TX power issue (see https://github.com/micropython/micropython-lib/pull/572#issuecomment-1529490736)~~ - Root cause remains unknown, have noted it in the README and the driver source.

 This work was funded through GitHub Sponsors.

mattytrentini commented:
Would be keen to see if we could use this driver on the Seeed E5 range - particularly the [Mini](https://wiki.seeedstudio.com/LoRa_E5_mini/) and [Development Kit](https://wiki.seeedstudio.com/LoRa_E5_Dev_Board/). I have one of each, so I guess I should test them!

I suspect it will work just fine to control the modem - the E5's use a module housing a variant of the WL5 - but getting MicroPython running on those chips is a little painful (BOOT0 is not exposed so an ST-Link is required - see [9130](https://github.com/micropython/micropython/pull/9130)).

@andrewleech I expect you'll be interested in this too.

dpgeorge commented:
As per #501, we need to think how to structure the `micropython/lora` directory, and what packages need to be made to hold the different components.

projectgus commented:
@mattytrentini I was looking at the E5s for a ultra low power + long range project as well! You've inspired me to order one of their devkits.

I think it should generally work, the other fiddly part (for pretty all these modules) seems to be how the TCXO and any antenna switches are wired in, which is either board- or module- specific. Everything should be exposed in the modem constructor, but knowing what to put there generally requires some schematic spelunking!

I hadn't noticed the BOOT0 thing either, that's mildly annoying. I see they have implemented a DFU mode for the factory AT firmware, but looks like some custom UART protocol(!) and they've enabled RDP on the factory AT firmware. Sigh.

tcpipchip commented:
I would like to test the micropython too...
I have here the LOM204 and LSM110 modules.
STM32 sx1276 and sx1262
something already in repository ?

projectgus commented:
@tcpipchip I think these can probably work:

For LSM110A, I found a [datasheet here](https://github.com/SeongJiIoT/LSM1x0A_SDK_LoRaWAN104/blob/main/Document/DS_LSM110A_R02_221101.pdf). Is this the same module? It says it is based on STM32WLE5CC, so if you are happy to overwrite the factory SigFox firmware then the code from this PR (and the linked PRs) may be enough to make it work. However, there is an RF switch shown in the block diagram:

![Block Diagram with RF switch highlighted](https://user-images.githubusercontent.com/205573/227352398-80b0f33c-9a78-44c1-ab10-0d9632219f37.png)

This might need some different code to switch correctly for RX/TX modes. I couldn't find any more information about it in the datasheet or app note, though - for example which control pins it is connected to.

For LOM204, I can't find any official resources at all. Do you have a link? I saw some mention of STM32L072+SX1276, which is maybe similar to the CMWX1ZZABZ module on the B-L072Z-LRWAN1 dev board. There's a [PR for that dev board here](https://github.com/micropython/micropython/pull/10640) that might be useful.

tcpipchip commented:
First step got
![image](https://user-images.githubusercontent.com/11488460/227515619-10739eb8-0dba-4a47-b331-8ddd27a61bdf.png)

tcpipchip commented:
here the LSM110A :)
![image](https://user-images.githubusercontent.com/11488460/227546027-63182278-463d-46c7-a466-a9cdf568ead0.png)

projectgus commented:
@tcpipchip Great! The branch for this PR has actually fallen a bit behind the "base" PR, which I just updated.

If you can wait another week or so then I'll be updating this branch and re-testing on STM32WL5, and I can post some instructions then.

projectgus commented:
@tcpipchip Oh, you'll also need to build MicroPython from this open PR branch: https://github.com/micropython/micropython/pull/9130 - in order to get the extra core-layer driver for the SUBGHZ interface. The current master branch doesn't have this yet.

tcpipchip commented:
@projectgus 
Yes, of course, i can wait for another week!

tcpipchip commented:
i am doing a draft
https://lom204-cli-wisol.blogspot.com/2023/04/micropython-no-lsm110a-lora.html

projectgus commented:
This PR has been delayed a little because the NUCLEO-WL55JC1 transmit performance is poor on the High Power antenna, and I can't find out why. The Low Power antenna transmit performance is better at every power level.

I installed the STM32CubeWL drivers and find the power levels are as expected when using ST's example app, so I think either a register write or an antenna switch function is missing here. I just can't find the difference, yet.

---

> wow! can you instruct me on git to get this branch ?

@tcpipchip Thanks for being patient while I worked back around to this.

The easiest way to get the PR branch from MicroPython is to use the GitHub CLI ([docs here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally?tool=cli)). Once this is installed, if you change to your micropython directory (as cloned from GitHub) then run `gh pr checkout 9130`.

Then change to subdirectory `lib/micropython-lib` and run `gh pr checkout 572` to get this micropython-lib PR as well.

Then before building the `stm32` port for your board, edit the mpconfigboard.mk file for your board and add a manifest file:

```
$ diff --git a/ports/stm32/boards/NUCLEO_WL55/mpconfigboard.mk b/ports/stm32/boards/NUCLEO_WL55/mpconfigboard.mk
index 210f3058c1..a49a91a8e1 100644
--- a/ports/stm32/boards/NUCLEO_WL55/mpconfigboard.mk
+++ b/ports/stm32/boards/NUCLEO_WL55/mpconfigboard.mk
@@ -10,4 +10,4 @@ MICROPY_VFS_FAT = 0
 MICROPY_VFS_LFS2 = 1
 
 # Don't include default frozen modules because MCU is tight on flash space
-FROZEN_MANIFEST ?=
+FROZEN_MANIFEST ?= $(BOARD_DIR)/manifest.py
```

(This is in the directory for the NUCLEO_WL55 board, if you build firmware for a different board then edit the file in that directory instead.)

Finally create the [board manifest file](https://docs.micropython.org/en/latest/reference/manifest.html#manifest) to "freeze" the LoRa driver modules. Mine is `ports/stm32/boards/NUCLEO_WL55/manifest.py` and contents are:

```py
include("$(MPY_DIR)/extmod/uasyncio")

require("lora-async")
require("lora-stm32wl5")
```

Finally, build and flash the new firmware for your board. If you can run `import lora` from the REPL then it worked out!

This takes some time and patience to get right, it will be easier once more of the PRs have merged.

If you have any hurdles building the firmware, please ask for help on the [Discord Server](https://micropython.org/discord) or open a [Discussion thread](https://github.com/micropython/micropython/discussions) as it's not directly related to this PR.

projectgus commented:
> wow! can you instruct me on git to get this branch ?

@tcpipchip Thanks for being patient while I worked back around to this.

The easiest way to get the PR branch from MicroPython is to use the GitHub CLI ([docs here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally?tool=cli)). Once this is installed, if you change to your micropython directory (as cloned from GitHub) then run `gh pr checkout 9130`.

Then change to subdirectory `lib/micropython-lib` and run `gh pr checkout 572` to get this micropython-lib PR as well.

Then before building the `stm32` port for your board, edit the mpconfigboard.mk file for your board and add a manifest file:

```
$ diff --git a/ports/stm32/boards/NUCLEO_WL55/mpconfigboard.mk b/ports/stm32/boards/NUCLEO_WL55/mpconfigboard.mk
index 210f3058c1..a49a91a8e1 100644
--- a/ports/stm32/boards/NUCLEO_WL55/mpconfigboard.mk
+++ b/ports/stm32/boards/NUCLEO_WL55/mpconfigboard.mk
@@ -10,4 +10,4 @@ MICROPY_VFS_FAT = 0
 MICROPY_VFS_LFS2 = 1
 
 # Don't include default frozen modules because MCU is tight on flash space
-FROZEN_MANIFEST ?=
+FROZEN_MANIFEST ?= $(BOARD_DIR)/manifest.py
```

(This is in the directory for the NUCLEO_WL55 board, if you build firmware for a different board then edit the file in that directory instead.)

Finally create the [board manifest file](https://docs.micropython.org/en/latest/reference/manifest.html#manifest) to "freeze" the LoRa driver modules. Mine is `ports/stm32/boards/NUCLEO_WL55/manifest.py` and contents are:

```py
include("$(MPY_DIR)/extmod/uasyncio")

require("lora-async")
require("lora-stm32wl5")
```

Finally, build and flash the new firmware for your board. If you can run `import lora` from the REPL then it worked out!

This takes some time and patience to get right, it will be easier once more of the PRs have merged.

If you have any hurdles building the firmware, please ask for help on the [Discord Server](https://micropython.org/discord) or open a [Discussion thread](https://github.com/micropython/micropython/discussions) as it's not directly related to this PR.

tcpipchip commented:
hi, i am back
where arey are you testing ? VM, Real Linux or Windows ?
At moment i am tryng on Windows (gitcli)

got this 
C:\Users\Usuario\Desktop>git clone https://github.com/micropython
Cloning into 'micropython'...
remote: Not Found
fatal: repository 'https://github.com/micropython/' not found

C:\Users\Usuario\Desktop>git clone https://github.com/micropython/micropython
Cloning into 'micropython'...
remote: Enumerating objects: 109173, done.
remote: Counting objects: 100% (36552/36552), done.
remote: Compressing objects: 100% (3588/3588), done.
remote: Total 109173 (delta 34693), reused 33057 (delta 32962), pack-reused 72621Receiving objects: 100% (109173/109173)Receiving objects: 100% (109173/109173), 55.84 MiB | 4.90 MiB/s, done.

Resolving deltas: 100% (81692/81692), done.
Updating files: 100% (5179/5179), done.

C:\Users\Usuario\Desktop>cd micropython

C:\Users\Usuario\Desktop\micropython>gh pr checkout 9130
remote: Enumerating objects: 34, done.
remote: Counting objects: 100% (29/29), done.
remote: Total 34 (delta 29), reused 29 (delta 29), pack-reused 5
Unpacking objects: 100% (34/34), 22.99 KiB | 148.00 KiB/s, done.
From https://github.com/micropython/micropython
 * [new ref]             refs/pull/9130/head -> feature/stm32wl5x-subghz
Switched to branch 'feature/stm32wl5x-subghz'

C:\Users\Usuario\Desktop\micropython>cd lib

C:\Users\Usuario\Desktop\micropython\lib>cd micropython-lib

C:\Users\Usuario\Desktop\micropython\lib\micropython-lib>dir
 O volume na unidade C não tem nome.
 O Número de Série do Volume é 42BE-2A44

 Pasta de C:\Users\Usuario\Desktop\micropython\lib\micropython-lib

04/06/2023  07:54    <DIR>          .
04/06/2023  07:54    <DIR>          ..
               0 arquivo(s)              0 bytes
               2 pasta(s)   92.391.051.264 bytes disponíveis

C:\Users\Usuario\Desktop\micropython\lib\micropython-lib>gh pr checkout 572
remote: Enumerating objects: 1, done.
remote: Total 1 (delta 0), reused 0 (delta 0), pack-reused 1
Unpacking objects: 100% (1/1), 232 bytes | 29.00 KiB/s, done.
From https://github.com/micropython/micropython
 * [new ref]             refs/pull/572/head -> objgenerator-exc_state
Deletion of directory 'lib/micropython-lib' failed. Should I try again? (y/n)

![image](https://github.com/micropython/micropython-lib/assets/11488460/04046116-b97e-46c1-bce6-c64aca1ad25d)

tcpipchip commented:
i changed the windows permission but got the same error...

Can you try there on windows ?

projectgus commented:
@tcpipchip I don't have a Windows system available to build MicroPython, and I'm afraid that I don't think building "stm32" port on Windows is supported apart from under WSL. You might want to try either WSL or a VM.

For the exact problem you're seeing now, try starting with `git clone --recursive https://github.com/micropython/micropython.git` to get a local copy that includes the "git submodules". Then you should find that the lib/micropython-lib directory is not empty, and the other steps you ran should be correct.

However, I think there will probably be other problems after this point. As building under Windows is not related to the code changes in this PR, please ask follow-up questions on [Discord](https://discord.com/channels/574275045187125269) or by starting a new thread in the [Discussions](https://github.com/micropython/micropython/discussions). You can @-mention me in either place if you like, I'll try to help more although I'm not very knowledgeable about Windows so probably I won't be much help with that.

tcpipchip commented:
Hi, got good progress...the --recursive did a big difference! :)
![image](https://github.com/micropython/micropython-lib/assets/11488460/8bb86274-fa1d-4930-bb5d-255d79a8ad9f)
what is the next procedure ?

tcpipchip commented:
![image](https://github.com/micropython/micropython-lib/assets/11488460/7c2d5b5e-b4cc-4fc2-9ebe-68b6fa2b43c3)

The problem is that is not mouting the FS

Can you help me ?

In paralell, i am doing the blog draft

tcpipchip commented:
i sent too this question to Micropython group...

tcpipchip commented:
no sucess on Forum HELP

andrewleech commented:
@tcpipchip where did you ask? This problem doesn't sound related to this PR at all, it would be good to try to keep the discussion here on topic :-) That being said you might just need to format the flash https://docs.micropython.org/en/latest/reference/filesystem.html#littlefs

projectgus commented:
@tcpipchip I've replied to you on the Discussions, here:
https://github.com/orgs/micropython/discussions/11708#discussioncomment-6118083

> no sucess on Forum HELP

Please refrain from this kind of reply, especially after only 48 hours. I know you're anxious to get this working, but I have other demands on my time. I see your posts when you make them, and I reply when I have time to do so. Hassling me in addition to that isn't helpful.

projectgus commented:
Rebased now that `lora`  support is in master, updated to use the new hardware interface from the `micropython` PR.

Keeping this in Draft until I get a chance to figure out the HP TX power issue, though.

projectgus commented:
After discussing with @dpgeorge I've called off the bug hunt on the HP antenna power ouput for now, and have noted it as a known issue in the README and the code.

Hoping it's solved sooner or later!

dpgeorge commented:
@projectgus can you please rebase on latest master, now that #712 is merged.

dpgeorge commented:
Also, please include appropriate version bumps in the manifest.py of those existing modules that have changed.

projectgus commented:
@dpgeorge Done, and thanks for bumping the manifest.py versions in those other PRs. Will note it from now on!

Have pushed another commit here to fix the busy period timeout handling, it was manifesting as a sporadic BUSY timeout on STM32WL5 the first time the modem initialised after power on reset.

projectgus commented on micropython/lora/lora-stm32wl5/lora/stm32wl5.py:
```diff
@@ -0,0 +1,136 @@
+# MIT license; Copyright (c) 2022 Angus Gratton
+#
+# This driver is essentially an embedded SX1262 with a custom internal interface block.
+# Requires the machine.SubGHzRadio() support in MicroPython to access.
```
Good catch, thanks. Have updated!

dpgeorge commented:
Thanks for all the work on this @projectgus !

tcpipchip commented:
that´s nice

tcpipchip commented:
Then now lorawan is part of micropython GitHub, right ?

projectgus commented:
> Then now lorawan is part of micropython GitHub, right ?

Kind of! It's a little complex.

The lora driver is merged and now has STM32WL5 support, so you can send/receive LoRa messages from this SoC using the github master branch version of micropython & micropython-lib.

The stm32wl5 build still doesn't include the Python lora driver by default, you'll have to [freeze the lora modules in](https://docs.micropython.org/en/latest/reference/packages.html#freezing-packages) which requires building the firmware. This might change soon, no we have all the support merged!

LoraWAN is an extra layer on top of LoRa to send packets with encryption, talk to gateways, etc. Micropython doesn't have an official driver for this yet.
